---
title: "Microbial Mapping - 16S NMDS"
output:
  html_document:
    df_print: paged
---
# Document Summary
This code is for generating figures related to the actual sequencing data from the waimea hiseq data sets.
first and foremost, the goal is to generate a variety of nmds plots to visualize braod trends in the data.


## Set Up Environment
Set a standardized color palettes for plotting and load libraries

```{r}
library(dplyr)
library(ggplot2)
library(phyloseq)
library(data.table)
library(parallel)
library(viridisLite)

set.seed(seed = 2020)

# set cores for parallel tasks
if( is.na( detectCores() )  ){
  no_cores <- 7
} else{
  no_cores <- detectCores()
}

# set working directory
if(getwd() != "/home/sean/Documents/Bioinformatics/Projects/Waimea/microbial_mapping/main")
  {
setwd("/home/sean/Documents/Bioinformatics/Projects/Waimea/microbial_mapping/main")
}


# set ggplot theme
theme_set(theme_bw())

# set color palettes 
pal_d = "Dark2"
scale_colour_discrete <-  function(palname=pal_d, ...){
  scale_colour_brewer(palette=palname, ...)
}
scale_fill_discrete <-  function(palname=pal_d, ...){
    scale_fill_brewer(palette=palname, ...)
}

pal_c = "c"
uniform_gradient_color <- function(palname = pal_c){
  scale_color_viridis_c(option = palname)
}

# can change these settings with viridis package
# with colour brewer, need add a line to each plot for continuous colors
# "scale_color_gradient2(low = "green", mid = "blue", high = "red")"

#  scale_colour_continuous <-  function(palname=pal_c, ...){
#    scale_colour_brewer(palette=palname, ...)
#  }
# # 
#  scale_fill_continuous <-  function(palname=pal_c, ...){
#    scale_fill_brewer(palette=palname, ...)
#  }



```


## Read in phyloseq object
```{r}
MM_16S_phy <- readRDS("../data/processed/cleaned/MM_16S_phy.rds")

MM_16S_phy

meta <- MM_16S_phy@sam_data %>% as.matrix %>% as.data.table


```




## all sample nmds
run nmds on all samples

### Ordinate

```{r}
# ordinate all phyloseq object
MM.ord <- ordinate(physeq = MM_16S_phy,
                   method =  "NMDS",
                   distance = "bray")
```

### Plot Ordinations

```{r}

# plot by sample and habitat
p <- plot_ordination(MM_16S_phy, MM.ord, type="samples", color = "habitat", shape = "host")
p <- p + scale_shape_discrete(solid = T)
p

ggsave("../outputs/figures/nmds/all_samples_habitat_nmds.png")


p <- plot_ordination(MM_16S_phy, MM.ord, type="samples", color = "trophic", shape = "host")
p <- p + scale_shape_discrete(solid = T) 
p

ggsave("../outputs/figures/nmds/all_samples_trophic_nmds.png")


```


### Just Host Associated
```{r}
MM  <- subset_samples(MM_16S_phy,
                      host != "Nonhost")

MM  <- prune_taxa(taxa_sums(MM) > 0, MM)



# ordinate
MM.ord <- ordinate(MM,
                   "NMDS",
                   distance = "bray")

# by sample
p = plot_ordination(MM, MM.ord, type="samples", color = "habitat", shape = "host", title="NMDS, Bray , Transect Samples, All OTUs")
p + scale_shape_discrete(solid = T)
rm(list=c("MM","MM.ord"))
gc()

```


### by trophic
split up data by habitat and run nmds colored by trophic level for all samples

```{r}

# list of unique habitats
habitat_types <- unique(MM_16S_phy@sam_data$habitat)

# function to subset and ordinate each habitat type
ordinate_habitat <- function(x){
  ## TEST
  ## x = habitat_types[1]
 
  # superassign
  hab <<- x
  
  # subset and ordinate
  MM_sub <- subset_samples(MM_16S_phy, habitat == hab)
  MM.ord <- ordinate(MM_sub, "NMDS", "bray")
  
  #plot
  p <- plot_ordination(MM_sub, MM.ord, type="samples", color = "trophic", shape = "host", title= paste("NMDS, Bray,", hab))
  p <- p +
    scale_shape_discrete(solid = T)
  return(p)
}

# run function
habitat_plots <- lapply(habitat_types, ordinate_habitat)

# display plots
habitat_plots

# save plots
names(habitat_plots) <- habitat_types

for(i in names(habitat_plots)){
  ggsave(filename = paste("../outputs/figures/nmds/", i,"_trophic_nmds.png", sep = ""), plot = habitat_plots[[i]])
}



```

### by habitat
```{r}

```

### by host
```{r}

```

## subset nmds
```{r}

```
