---
title: "Microbial Mapping - 16S NMDS"
output:
  html_document:
    df_print: paged
---
# Document Summary
This code is for generating figures related to the actual sequencing data from the waimea hiseq data sets.
first and foremost, the goal is to generate a variety of nmds plots to visualize braod trends in the data.


## Set Up Environment
Set a standardized color palettes for plotting and load libraries

```{r}
library(dplyr)
library(ggplot2)
library(phyloseq)
library(data.table)
library(parallel)
library(viridisLite)

set.seed(seed = 2020)

# set cores for parallel tasks
if( is.na( detectCores() )  ){
  no_cores <- 7
} else{
  no_cores <- detectCores()
}

# set working directory
# if(getwd() != "/home/sean/Documents/Bioinformatics/Projects/Waimea/microbial_mapping/main")
#   {
# setwd("/home/sean/Documents/Bioinformatics/Projects/Waimea/microbial_mapping/main")
# }


# set ggplot theme
theme_set(theme_bw())

# set color palettes 
pal_d = "Dark2"
scale_colour_discrete <-  function(palname=pal_d, ...){
  scale_colour_brewer(palette=palname, ...)
}
scale_fill_discrete <-  function(palname=pal_d, ...){
    scale_fill_brewer(palette=palname, ...)
}

pal_c = "plasma"
uniform_gradient_color <- function(palname = pal_c){
  scale_color_viridis_c(option = palname)
}

# can change these settings with viridis package
# with colour brewer, need add a line to each plot for continuous colors
# "scale_color_gradient2(low = "green", mid = "blue", high = "red")"

#  scale_colour_continuous <-  function(palname=pal_c, ...){
#    scale_colour_brewer(palette=palname, ...)
#  }
# # 
#  scale_fill_continuous <-  function(palname=pal_c, ...){
#    scale_fill_brewer(palette=palname, ...)
#  }



```


## Read in phyloseq object
```{r}
MM_16S_phy <- readRDS("../data/processed/cleaned/MM_16S_phy.rds")

MM_16S_phy

meta <- MM_16S_phy@sam_data %>% as.matrix %>% as.data.table


sample_number_by_site <- meta[, .N, by = .(site_name, habitat)][order(site_name, habitat)]
fwrite(sample_number_by_site, "../data/processed/interim/sample_number_by_site.csv")


```




## all sample nmds
run nmds on all samples, except controls

### Ordinate

```{r}
# subset to remove controls (no metadata in the host column)
nsamples(MM_16S_phy)

MM_16S_phy <- subset_samples(MM_16S_phy, host != "")


# ordinate all phyloseq object

MM.ord <- ordinate(physeq = MM_16S_phy,
                   method =  "NMDS",
                   distance = "bray")
```

### Plot Ordinations

```{r}

# plot by sample and habitat
p <- plot_ordination(MM_16S_phy, MM.ord, type="samples", color = "habitat", shape = "host")
p <- p + scale_shape_discrete(solid = T)+
      ggtitle("All Samples, Bray Curtis Distance, NMDS - By Habitat")
p

ggsave("../outputs/figures/nmds/all_samples_habitat_nmds.pdf")


p <- plot_ordination(MM_16S_phy, MM.ord, type="samples", color = "trophic", shape = "host")
p <- p + scale_shape_discrete(solid = T)  +
      ggtitle("All Samples, Bray Curtis Distance, NMDS - By Trophic")
p

ggsave("../outputs/figures/nmds/all_samples_trophic_nmds.pdf")


p <- plot_ordination(MM_16S_phy, MM.ord, type="samples", color = "elevation", shape = "host") +
     scale_shape_discrete(solid = T)  +
     uniform_gradient_color()  +
      ggtitle("All Samples, Bray Curtis Distance, NMDS - By Elevation")
p

ggsave("../outputs/figures/nmds/all_samples_elevation_nmds.pdf")


```


### Just Host Associated
```{r}
MM  <- subset_samples(MM_16S_phy,
                      host != "Nonhost")

MM  <- prune_taxa(taxa_sums(MM) > 0, MM)



# ordinate
MM.ord <- ordinate(MM,
                   "NMDS",
                   distance = "bray")

# by sample
p = plot_ordination(MM, MM.ord, type="samples", color = "habitat", shape = "host", title="Host-Associated Samples, NMDS, Bray - By Habitat ")
p + scale_shape_discrete(solid = T)


ggsave("../outputs/figures/nmds/all_host_associated_nmds.pdf")

rm(list=c("MM","MM.ord"))
gc()

```

## Subsets
### by habitat
split up data by habitat and run nmds colored by trophic level for all samples

```{r}

# list of unique habitats
habitat_types <- unique(MM_16S_phy@sam_data$habitat)

# function to subset and ordinate each habitat type
ordinate_habitat <- function(x){
  ## TEST
  ## x = habitat_types[1]
 
  # superassign
  hab <<- x
  
  # subset and ordinate
  MM_sub <- subset_samples(MM_16S_phy, habitat == hab)
  MM.ord <- ordinate(MM_sub, "NMDS", "bray")
  
  #plot
  p <- plot_ordination(MM_sub, MM.ord, type="samples", color = "trophic", shape = "host", title= paste("NMDS, Bray,", hab))
  p <- p +
    scale_shape_discrete(solid = T)
  return(p)
}

# run function
habitat_plots <- lapply(habitat_types, ordinate_habitat)

# display plots
habitat_plots

# save plots
names(habitat_plots) <- habitat_types

for(i in names(habitat_plots)){
  ggsave(filename = paste("../outputs/figures/nmds/habitat", i,"_nmds.pdf", sep = ""), plot = habitat_plots[[i]])
}



```

### by trophic level
```{r}

# list of unique habitats
trophic_types <- unique(MM_16S_phy@sam_data$trophic)

# function to subset and ordinate each habitat type
ordinate_troph <- function(x){
  ## TEST
 
  # superassign
  troph <<- x
  
  # subset and ordinate
  MM_sub <- subset_samples(MM_16S_phy, trophic == troph)
  MM.ord <- ordinate(MM_sub, "NMDS", "bray")
  
  #plot
  p <- plot_ordination(MM_sub, MM.ord, type="samples", color = "elevation", shape = "habitat", title= paste("NMDS, Bray,", troph))  +
    uniform_gradient_color()+
    scale_shape_discrete(solid = T)
  return(p)
}

# run function
trophic_plots <- lapply(trophic_types, ordinate_troph)

# display plots
trophic_plots

# save plots
names(trophic_plots) <- trophic_types

for(i in names(trophic_plots)){
  ggsave(filename = paste("../outputs/figures/nmds/trpohic_", i,"_nmds.pdf", sep = ""), plot = trophic_plots[[i]])
}

```

### by host
```{r}

```

## subset nmds
```{r}

```
